### ISSUE TITLE
RPC response takes ~25-35 seconds

### ISSUE THREAD

**Issue by nicholasdkunes** [justinfx/pyRpc#7]
Server:

```
def insa_dispatch_rpc():
	print(insa_c.magenta + "$insa> rpc triggered" + insa_c.white)
	return "heck"

def rpc_serve():

	globals.insa_rpc_server = PyRpc("insa_analytics", tcpaddr="127.0.0.1:4000") 

	globals.insa_rpc_server.publishService(insa_dispatch_rpc)

	globals.insa_rpc_server.start()

	print(insa_c.magenta + "$insa> rpc server is online" + insa_c.white)

	while True:
		time.sleep(.1)

def start_rpc_server():
	threading.Thread(target = rpc_serve, args = ()).start()
```

Client:

```
def retrieve_general_data(request):

	remote = RpcConnection("insa_analytics", tcpaddr="127.0.0.1:4000")

	_result = None
	_result = remote.call("insa_dispatch_rpc")

	remote.close()
	return HttpResponse(str(json.dumps(str(_result.result))))

```

retrieve_general_data is called instantly, I have tested this by returning an HttpResponse with a string. It is called instantaneously. For remote.call(...) to return, it takes around 25-35 seconds. Why is this? I tried raising workers to 32 threads. 

Also, without specifying tcpaddr, it never connects to begin with. Yes, they are on the same host system.

**Comment by justinfx** [comment]
That does sound like an extremely long time to respond, in a call that doesn't do any work other than a print function. Does this happen only under heavy load or even with an idle server and a single request? I wouldn't think raising workers to 32 would be relevant unless this only happens under heavy load and with a service function that does real work.

Can you run the client and server with debug logging enabled? If you aren't currently using the python logger, just add this in both your server and client main:

```
import logging
logging.basicConfig(level=logging.DEBUG)
```

I would like to know where the delays are coming from. Have you identified if it is blocking at the client side or the server side? Can you reproduce this even in a simpler client that just starts up, makes the request, and quits?

> Also, without specifying tcpaddr, it never connects to begin with. Yes, they are on the same host system.

That is surprising, seeing as the default when not specifying a tcpaddr is to use inter-process communication over a named socket. What OS are you running on? It could have something to do with your zmq version and your OS. I wrote this simple RPC solution 11 years ago and haven't been tracking zmq for a very long time.

**Comment by nicholasdkunes** [comment]
> I would like to know where the delays are coming from. Have you identified if it is blocking at the client side or the server side? Can you reproduce this even in a simpler client that just starts up, makes the request, and quits?

After some debugging, it is blocking on the client as far as I can tell. This call is taking all the time:
```remote = RpcConnection("insa_analytics", tcpaddr="127.0.0.1:4000")```

Version Info:
```Distributor ID: Ubuntu
Description:    Ubuntu 20.04.4 LTS
Release:        20.04
Codename:       focal
```

pyzmq:
```Name: pyzmq
Version: 23.1.0
Summary: Python bindings for 0MQ
Home-page: https://pyzmq.readthedocs.org
Author: Brian E. Granger, Min Ragan-Kelley
Author-email: zeromq-dev@lists.zeromq.org
License: LGPL+BSD
Location: /usr/local/lib/python3.8/dist-packages
Requires:
Required-by: pyRpc
```

**Comment by justinfx** [comment]
I just ran the existing [server.py](/justinfx/pyRpc/blob/master/examples/server.py) and [client.py](/justinfx/pyRpc/blob/master/examples/client.py) examples, with `tcpaddr="127.0.0.1:4000"`, using python 3.8.10 on ubuntu 20.04.  This example code does not replicate the delayed client connection issue you are reporting.

I then modified the client code to construct and close a new `RpcConnection` on each request. While I am not able to reproduce the exact 30 second delay when constructing a new client `RpcConnection` I did notice an error when trying to construct the second connection after the first one was closed:

```
zmq.error.ZMQError: Address already in use
```

The reason this is happening is that by default an `RpcConnection` uses the shared global instance of `zmq.Context`, which is the thread-safe instance that manages the underlying network sockets. When you call close on the connection without having passed in your own context, you are effectively closing the global context. Now in my case that creates an error, and I am hoping it would be the same cause of your issue.

The fix I recommend is to remove the call to `close()` at the end of your handler. According to zmq, it is safe to share the global Context instance, and to then have different threads construct their own socket connection around that, which is what `RpcConnection` is doing. Given that you are trying to implement a web handler, I think this might be the most appropriate approach.
